### PR TITLE
Again, but in Python

### PR DESCRIPTION
**Pull request by christopher-dG**
Will probably be easier to maintain this way. Tested on a registry fork and it's working identically.

### CODE REVIEW

**Comment by christopher-dG** [comment]
Anyone know why CI fails? Supposedly `pytest` doesn't find any tests, however it works locally...

**Comment by fchorney** [comment]
> Anyone know why CI fails? Supposedly `pytest` doesn't find any tests, however it works locally...

How are you running pytest locally?

**Comment by christopher-dG** [comment]
```sh
$ pytest
```

My environment looks like this (in a virtualenv):

```
appdirs==1.4.3
atomicwrites==1.3.0
attrs==19.1.0
black==19.3b0
certifi==2019.3.9
chardet==3.0.4
Click==7.0
idna==2.8
importlib-metadata==0.18
more-itertools==7.0.0
packaging==19.0
pluggy==0.12.0
py==1.8.0
pyparsing==2.4.0
pytest==4.6.3
python-gitlab==1.8.0
requests==2.22.0
six==1.12.0
toml==0.10.0
urllib3==1.25.3
wcwidth==0.1.7
zipp==0.5.1
```

I did notice that my local pytest is newer than what's installed on the CI box.

**Comment by fchorney** [comment]
when I make a virtualenv and then install the requirements and run `pytest` I just get

```
$ pytest
=========================================================== test session starts ===========================================================
platform darwin -- Python 3.7.2, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
rootdir: /Users/fchorney/Documents/repos/invenia/TagBotGitLab
collected 0 items / 1 errors

================================================================= ERRORS ==================================================================
_____________________________________________________ ERROR collecting test_tagbot.py _____________________________________________________
test_tagbot.py:2: in <module>
    import tagbot
tagbot.py:12: in <module>
    registrator = int(os.environ["REGISTRATOR_ID"])
venv/lib/python3.7/os.py:678: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'REGISTRATOR_ID'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================================= 1 error in 0.26 seconds =========================================================
```

**Comment by fchorney** [comment]
Also just curious, whats the `package.json` and `package-lock.json` for?

**Comment by christopher-dG** [comment]
Oh shoot, yeah it's the environment variables. Good catch!
The `package*` stuff is for the [Python requirements plugin](https://github.com/UnitedIncome/serverless-python-requirements).

**Comment by christopher-dG** [comment]
Welp, I could have sworn that I read the CI logs earlier and didn't see that error message but it's right there in the open...

File: tagbot.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import gitlab
```
Comment (fchorney) [inline tagbot.py]: Just some style stuff. 
According to our iSort rules `import gitlab` should come after the python stuff

```python
import json
import os
import re
import traceback

import gitlab


re_repo = re.compile("Repository:.*/(.*/.*)")
```

File: tagbot.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        return None, None, None, "No version match"
+    version = m[1].strip()
+    m = re_commit.search(body)
+    if not m:
+        return None, None, None, "No commit match"
+    commit = m[1].strip()
+    return repo, version, commit, None
+
+
+def get_in(d, *keys, default=None):
```
Comment (fchorney) [inline tagbot.py]: I'm not necessarily opposed to this function, but it feels a little unnecessary for this script, imo.

Comment (christopher-dG) [reply tagbot.py]: I agree it's not strictly necessary, I just like to not worry about `KeyError`s and dislike `d.get(k1, {}).get(k2, {})`. It's fairly simple and tested so I'd prefer for it to stay.

File: tagbot.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+)
+
+
+def handler(evt, _ctx):
+    """Lambda entrypoint."""
+    try:
+        if get_in(evt, "headers", "X-Gitlab-Token") != token:
+            status, err = 403, "Invalid token"
+        else:
+            status, err = 200, handle_event(json.loads(evt.get("body", "{}")))
```
Comment (fchorney) [inline tagbot.py]: Are there possibly exceptions we could `raise` in other functions that bubble up here to give a little more insight on where something failed? Could raise `RuntimeError`s in the other functions when something bad happens, and then the error messages would bubble up here.

Comment (fchorney) [reply tagbot.py]: Or at least put the actual error message that gets raised in in the `err` error text?

File: tagbot.py (hunk 2 above)
Comment (fchorney) [reply tagbot.py]: hmm yeah I guess it handles key errors. aight fair enough.

File: tagbot.py (hunk 3 above)
Comment (christopher-dG) [reply tagbot.py]: There are some exceptions that can be raised, yes. The way it currently works is basically:

- Any "validation" error goes to `err` as a simple message i.e. "this isn't a MR" and gets printed
- Any actual exception gets caught and printed via `traceback.print_exc()`

So any error that gets raised still does end up in the Cloudwatch logs, with a full stacktrace for exceptions.

Comment (fchorney) [reply tagbot.py]: OH, yeah sorry I missed that traceback.print_exc()

**Comment by fchorney** [comment]
I think it would be nice to add more tests that mock the gitlab responses and how the script deals with them, but aside from that I think this seems good.

**Comment by christopher-dG** [comment]
Task for a new PR perhaps?

**Comment by christopher-dG** [comment]
Actually, mocking looks quite simple so I'll spend a bit of time to get that done.

**Pull request merged**